sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala

```diff
@@ -866,71 +867,189 @@ class Analyzer(
    * Note: CTEs are handled in CTESubstitution.
    */
   object ResolveSubquery extends Rule[LogicalPlan] with PredicateHelper {
-
     /**
-     * Resolve the correlated predicates in the clauses (e.g. WHERE or HAVING) of a
-     * sub-query by using the plan the predicates should be correlated to.
+     * Resolve a subquery using the outer plan. This rule creates a dedicated analyzer which can
+     * also resolve outer plan references.
      */
-    private def resolveCorrelatedSubquery(
-        sub: LogicalPlan, outer: LogicalPlan,
-        aliases: scala.collection.mutable.Map[Attribute, Alias]): LogicalPlan = {
-      // First resolve as much of the sub-query as possible
-      val analyzed = execute(sub)
-      if (analyzed.resolved) {
-        analyzed
-      } else {
-        // Only resolve the lowest plan that is not resolved by outer plan, otherwise it could be
-        // resolved by itself
-        val resolvedByOuter = analyzed transformDown {
-          case q: LogicalPlan if q.childrenResolved && !q.resolved =>
-            q transformExpressions {
-              case u @ UnresolvedAttribute(nameParts) =>
-                withPosition(u) {
-                  try {
-                    val outerAttrOpt = outer.resolve(nameParts, resolver)
-                    if (outerAttrOpt.isDefined) {
-                      val outerAttr = outerAttrOpt.get
-                      if (q.inputSet.contains(outerAttr)) {
-                        // Got a conflict, create an alias for the attribute come from outer table
-                        val alias = Alias(outerAttr, outerAttr.toString)()
-                        val attr = alias.toAttribute
-                        aliases += attr -> alias
-                        attr
-                      } else {
-                        outerAttr
-                      }
-                    } else {
-                      u
-                    }
-                  } catch {
-                    case a: AnalysisException => u
-                  }
-                }
+    def apply(plan: LogicalPlan): LogicalPlan = plan resolveOperators {
+      // Only a few unary nodes (Project/Filter/Aggregate/Having) can contain subqueries.
+      case q: UnaryNode if q.childrenResolved =>
+        q transformExpressions {
+          case e: SubqueryExpression if !e.query.resolved =>
+            val analyzer = new Analyzer(catalog, conf) {
+              override val extendedCheckRules = self.extendedCheckRules
+              override val extendedResolutionRules = self.extendedResolutionRules :+
+                ResolveOuterReferences(q.child, resolver)
             }
+            e.withNewPlan(analyzer.execute(e.query))
```

Comment:
You example will not resolve without an outer reference :)...

I do get your point; this will go wrong as soon as you use unqualified names and stack two projections on top of each other. For example (assume the outer relation also contains an `a` reference):
```scala
Project('a, Project(Alias('b, "a")(), LocalRelation('b)))
```
I'll restore the functionality (lets use a loop instead recursion though).